### ISSUE TITLE
HTTP Status Code and Query Errors

### ISSUE THREAD

**Issue by nathanielc** [influxdata/influxdb#13594]
When should an error in processing a query return an non 200 HTTP status code and when should it not?

The current behavior is that once the query request passes some basic validation and starts executing an HTTP 200 status code is returned and any errors are directly encoded in the csv format. In other words there are two classes of errors in the query API.

* Compile time errors - these are errors that can be discovered by statically analysing the Flux script for any errors
* Runtime errors - these are errors that are only discovered once execution has begun and data is being read.

Compile time errors result in non 2XX HTTP status codes.
Runtime errors result in 200 HTTP status codes and are encoded into the csv response.

The runtime error return 200 HTTP status codes because we stream results to the client, once we start processing a query it is not know whether it will succeed or encounter an error. So the current behavior is to write out the HTTP 200 status code so that the client expects to get csv data, at which point the csv data may contain results and error or possibly both.

**Comment by rbetts** [comment]
Thinking about this ..

 (a) Part of the problem is that since we don't expose quorum read semantics to the end-user, it is hard to know what they want. 
 (b) We can't buffer until seeing the last byte in some cases (for example, returning an entire bucket).
 (c) Exposing partition ids isn't useful to the end-user; however, exposing a count of partitions available and unavailable might be more useful?
 (d) A partial read in an AP system isn't a failure - it's a desired feature.. but without a way for end-users to request quorum reads .. it is very hard to give all the info back that's needed.

**Comment by nhaugo** [comment]
I have two problems with this. First it makes it difficult to get visibility into the user experience and to do error tracking across the system.
Second, it removes meaning and context from http. Users can not rely on the status code as being a meaningful indicator. Users will need to consult the output of the query and handle all exceptions in the result handling step. Applications will have to always assume there could be an error in result, which makes for unpleasant application building.

**Comment by mark-rushakoff** [comment]
The workflow I had been using when writing the queries for the usage dashboards was typically to view the raw CSV data until I was happy with the returned data, at which point I'd get into visualization details.

It was a bad experience when the first several tables looked right and then one later either had a "partition unavailable" error, or there was a Flux runtime error, typically due to an unexpectedly missing tag causing an invalid `group` method call or similar.

Basically I would iterate one Flux method at a time on my query, and it was frustrating when I would do 

```
...
  |> a()
  |> b()
  |> c()
  |> d()
```

And then my entire query would fail on `d()`, but I would have to backtrack to `a()` to find out that one table at that point didn't work.

Maybe that workflow can be solved in the UI by more prominently calling out when a single table in a stream has an error.

**Comment by nhaugo** [comment]
> Part of the problem is that since we don't expose quorum read semantics to the end-user, it is hard to know what they want.
I don't agree that we don't know what they want. They want the data that was requested. If a partition (or any part of the system) is unavailable for and unexpected reason than an error has occurred and the system should clear about that. Users are only interested in the read semantics if they are trying to get to the bottom of an issue. I think that partial failures should be represented as such, even if we have additional information and debugging to give them.

CS theory shouldn't be the guiding principals of what is in this case a user experience issue.

**Comment by rbetts** [comment]
When I say that we "don't know what users want", I mean that we don't know if they want to wait for the first byte until the entire query computed or if they want streamed results.  Both are valid and useful policies; we need a way for users to specify which semantics they prefer. That is what C*, dynamo, elastic, and almost all other AP systems do. We should supply the capability and let the user choose the policy.

**Comment by chnn** [comment]
Might be worth thinking about this in the context of the rest of the InfluxDB API as well. Since Flux queries are proxied through the API, an API consumer  needs to consider all the ways a non-query API request can fail as well.

In the UI, that has led us to check for the following errors, roughly in this order:

1. A `error` column in the first table of a 200 Flux CSV response
2. A string under the `message` key in a 4xx or 5xx JSON response
3. A string under the `error` key in a 4xx or 5xx JSON response
4. A nonempty response body in a 4xx or 5xx response (e.g. a stacktrace from a panic)

If the response is still a 4xx or 5xx status code, we fall back to the HTTP status text (e.g. "Bad Request"). 

We currently don't check for these errors, though we probably should:

- Errors in the `X-Influx-Error` header
- Errors in anything but the first Flux table in the response 

The distinction between streaming / non-streaming errors make sense to me, but ideally there would only be one place to check for the error code and message of non-streaming error.
